docs/contrib/workflow.md

```diff
@@ -0,0 +1,232 @@
+# Contribution to OpenKilda project
+
+Anyone can contribute in OpenKilda project. This instruction describes how
+OpenKilda contribution looks like: from an idea to OpenKilda release.
+
+# Workflow
+
+## Idea
+
+If you are willing to add a new feature or make an improvement it makes sense
+to discuss it with other contributors first. To find current active contributors
+you can open a list of [open pull requests](https://github.com/telstra/open-kilda/pulls)
+and pick an author of recent pull request to write him/her a message about your
+idea. He or she will invite you to a place for idea discussions.
+
+If you don't have an idea, but you still want to contribute - please take a look on a
+list of [open issues](https://github.com/telstra/open-kilda/issues). If you find
+an open and unassigned issue (of course you will) you can assign it to yourself and
+start to work on it.
+
+### Design
+
+If you are working on a small feature, fix or improvement you can skip the design step.
+But if you are working on a big feature, especially which changes API, you must
+create a design document first. Examples of design documents can be found [here](../design).
+
+## Development
+
+### Git branches
+
+To make any changes in OpenKilda repository you need to create a git branch.
+There are 2 main branches in the repository:
+
+* `develop` - contains changes which were not yet released
+* `master` - contains last [release](https://github.com/telstra/open-kilda/releases)
+
+Please checkout your branch from `develop` branch.
+
+### Branch name
+
+There are no strict branch name restrictions. The only rule is to make branch name meaningful.
+In practice, we are using the following prefixes for branches:
+
+* `doc/` - for adding/updating the documentation
+* `feature/` - for adding new features
+* `fix/` - for fixing current codebase
+* `test/` - for tests
+* `release` - for releases (only a release manager can create such branches)
+* `hotfix` - for hotfixes which must be released ASAP (only a release manager can create such branches)
+
+After the prefix there must be a very short description of a changes.
+
+Good branch names:
+
+* `doc/contribution_guide`
+* `feature/ha_flow_create`
+* `fix/incorrect_latency_validation`
+* `test/ha_flow_functional_tests`
+* `release-1.2.3`
+* `hotfix-1.2.4`
+
+Bad branch names:
+
+* `guide`
+* `refactoring`
+* `added_tests`
+* `my_mega_fix`
+* `doc/how_to_contribute_in_kilda_if_you_cant_come_up_with_a_short_name_for_a_git_branch`
+
+### Commits
+
+The best practice is to have one commit per one branch. If you need to change
+your code after it was pushed - just amend your new changes into your last commit
```

Comment:
To summarize. Having one-commit branches or changing the commits from the past brings several problems in using git and doing code review. Having more commits allows cherry-pick, revert, or just understand better, or save some time (like no need for multiple people really review commit "fix check style"). On the other hand, rebasing with multiple commits leads to more conflicts resolution. Multiple commits introduce a noise and not informative in git blame or history. 
So a compromise I see could be: squash commits before putting `ready to merge`. And the guide would be: in your feature branch do whatever you believe and agreed with collaborators is best, one or more commits, rebasing, merging, whatever. But when the work is done, a dev, responsible for a particular branch, squashes all commits into a single one, puts a summarizing message into this commit, and puts a label `ready to merge`. This way, the release engineer have a single-commit branch to merge.
I tried to do this with one of the branches: `fix/4984-sub-flow-descriptions`, you can check whether it is a good compromise. 
Here are the steps how to do that inside a feature branch:

_update all branches, using your favorite way, or:_
```
git fetch
git switch develop
git pull
git switch FEATURE_BRANCH
git pull
```
_squash the commits. The reset command reads all commits and creates a new CL as not committed changes:_
```
git switch FEATURE_BRANCH
git reset --soft $(git merge-base develop HEAD)
git commit -m "A nice summary commit message that will remain in the develop branch"
```
_submit changes to remote. It breaks history, so other people are not supposed to work on this branch at this moment:_
```
git push --force
```
_Optionally, verify how many commits are there not counting the merges:_
```
$ git rev-list --no-merges --count HEAD ^develop
1
```
_In this example there is just one commit in the current branch._

Reply:
It's been almost a year. I've started to re-review the PR after the push --force and now it's difficult to understand the changes. I've tried comparing the commit id's but since we have introduced a lot of changes in the repository it's difficult to see the changes in the PR now.  I agree with @dmitrii-beliakov now. I think the best option is that just before the merge we PR owner must squash the commits into one (or two/three if the commits make sense by themselves).